rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,629 @@
+`tf.ModuleMetaclass` to wrap all methods in the modules name scope. This design
+allows for submodules or variables to created in any method (including
+`__init__` and `__call__`). There is more detail on this decision in the
+detailed design below.
+
+```python
+class Module(six.with_metaclass(ModuleMetaclass, tf.checkpoint.Checkpointable)):
+  """Base neural network module class.
+
+  A module is a named container for `tf.Variable`s, other `tf.Module`s and
+  functions which apply to user input. For example a dense layer in a neural
+  network might be implemented as a `tf.Module`:
+
+  >>> class Dense(tf.Module):
+  ...   def __init__(self, in_features, output_features):
+  ...     super(Dense, self).__init__()
+  ...     # NOTE: We're creating variables in the ctor, but you can actually
+  ...     #       create them in any method (including __call__) you like.
+  ...     self.w = tf.Variable(
+  ...         tf.random.normal([input_features, output_features]), name='w')
+  ...     self.b = tf.Variable(tf.zeros([output_features]), name='b')
+  ...
+  ...   def __call__(self, x):
+  ...     y = tf.matmul(x, self.w) + self.b
+  ...     return tf.nn.relu(y)
+
+  You can use the layer as you would expect:
+
+  >>> d = Dense(input_features=64, output_features=10)
+  >>> d(tf.ones([100, 64]))
+  <tf.Tensor: ...>
+
+  By subclassing `tf.Module` instead of `object` any variables created inside
+  the module are automatically created within the modules name scope:
+
+  >> d.w.name
+  "dense/w:0"
+
+  In eager mode this is useful for debugging, and when used with `@tf.function`
+  the use of name scopes gives operations (e.g. matmul) useful names as well.
+
+  As well as automatic naming, the Dense module inherits methods for tracking
+  its variables:
+
+  >>> d.variables
+  (<tf.Variable 'dense/b:0' ...>, <tf.Variable 'dense/w:0' ...>)
+  """
+
+  @property
+  def name(self):
+    """Returns the name of this module as passed or determined in the ctor.
+
+    NOTE: This is not the same as the `self.name_scope.name` which includes
+    parent module names.
+    """
+
+  @property
+  def name_scope(self):
+    """Returns a `tf.name_scope` instance for this class.
+
+    NOTE: The module name scope includes all parent modules as determined
+    when this module was created.
+    """
+
+  @property
+  def variables(self):
+    """Returns a collection of variables owned by this module and it's submodules.
+
+    Returns:
+      A collection of variables for the current module (sorted by attribute name)
+      followed by variables from all submodules recursively (depth first).
+    """
+
+  @property
+  def trainable_variables(self):
+    """Returns a collection of variables owned by this module and it's submodules.
+
+    Returns:
+      A collection of variables for the current module (sorted by attribute name)
+      followed by variables from all submodules recursively (depth first).
+    """
```

Comment:
I'm a bit late to the party, but I don't see this in the discussion anywhere:

Keras layers and models implement a `trainable` flag as well.
When set to False, none of the variables in the module, or submodules are returned in by trainable_variables.

The current implementation of module does not recognize a module-level trainable flag, it's not clear if that's intentional, or an omission.

Looking at the implementation... it would require a separate [predicate for recursion](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/module/module.py#L419).

This seems like a minor change that could reduce surprises and increase compatibility. Especially with the plan to make `layer` and `model` sub-classes of module.

WDYT?

Reply:
This is intentional. Module uses the immutable "training" bit on `tf.Variable` to determine which variables are trainable. If you want to implement layer freezing like Keras does for a `Module` subclass you can do:

```python
class KerasLike(tf.Module):

  def __init__(self, trainable=True, name=None):
    super(KerasLike, self).__init__(name=name)
    self.trainable = trainable

  @property
  def trainable_variables(self):
    if not self.trainable:
      return ()
    return super(KerasLike, self).trainable_variables
```

Reply:
Great. Thanks for the clarifications.